### PR TITLE
Story/tp 114353 setup sendmail action

### PR DESCRIPTION
**Pull request by sirolf2009**

### CODE REVIEW

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/autoconfiguration/SpotlerConfiguration.kt [hunk 1]
```diff
@@ -3,18 +3,27 @@ package com.ritense.spotler.autoconfiguration
...
     ) = SpotlerPluginFactory(
-        pluginService
+        pluginService,
+        restTemplate
     )
 
+    @Bean
+    @ConditionalOnMissingBean(RestTemplate::class)
+    fun restTemplate(): RestTemplate {
+        return RestTemplateBuilder().build()
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/autoconfiguration/SpotlerConfiguration.kt:27]: This is needs to have more that just a plain RestTemplate. Please include a timeout if the request takes to long.

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/config/FlowmailerProperties.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ritense.spotler.config
+
+import org.springframework.boot.context.properties.ConfigurationProperties
+
+@ConfigurationProperties(prefix = "valtimo.flowmailer")
+data class FlowmailerProperties (
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/config/FlowmailerProperties.kt:22]: FlowMailer?

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/Placeholder.kt
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/Placeholder.kt:1]: missing license

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/RecipientType.kt
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/RecipientType.kt:1]: missing license

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/SubmitMessage.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+package com.ritense.spotler.domain
+
+import com.ritense.valtimo.contract.mail.model.TemplatedMailMessage
+import org.apache.tika.Tika
+
+data class SubmitMessage(
+    val attachments: MutableList<Attachment> = mutableListOf(),
+    val data: MutableMap<String, Any>,
+    val flowSelector: String, //link to flow with template
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/SubmitMessage.kt:25]: flow selector is this still called this way with spotler?

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/plugin/SpotlerPlugin.kt [hunk 4]
```diff
@@ -22,14 +42,87 @@ class SpotlerPlugin {
...
+        @PluginActionProperty recipients: Array<SpotlerRecipient>,
+        @PluginActionProperty placeholders: Array<Placeholder>,
+        @PluginActionProperty mailTemplateIdentifier: String
+    ) {
+        val flowMailerProperties = FlowmailerProperties(
+            clientId,
+            clientSecret,
+            accountId
+        )
+        val flowmailerTokenService = FlowmailerTokenService(
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/plugin/SpotlerPlugin.kt:67]: Idem alot of Flow I see

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/service/FlowmailerDispatcher.kt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import org.springframework.http.HttpEntity
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpMethod
+import org.springframework.http.MediaType
+import org.springframework.web.client.HttpClientErrorException
+import org.springframework.web.client.HttpServerErrorException
+import org.springframework.web.client.HttpStatusCodeException
+import org.springframework.web.client.RestTemplate
+
+class FlowmailerMailDispatcher(
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/service/FlowmailerDispatcher.kt:37]: Flow...

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/service/FlowmailerTokenService.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.http.HttpEntity
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpMethod
+import org.springframework.http.MediaType
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.web.client.HttpClientErrorException
+import org.springframework.web.client.HttpStatusCodeException
+import org.springframework.web.client.RestTemplate
+
+class FlowmailerTokenService(
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/service/FlowmailerTokenService.kt:32]: Flow...

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/service/FlowmailerTokenService.kt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    fun <T> getType(responseClass: Class<out T>): ParameterizedTypeReference<T> {
+        val type: ParameterizedTypeReference<T> = ParameterizedTypeReference.forType(
+            ResolvableType.forClass(responseClass).type
+        )
+        return type
+    }
+
+    companion object {
+        private const val TOKEN_URL = "https://login.flowmailer.net/oauth/token"
```
Comment (Tom-Ritense) [inline spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/service/FlowmailerTokenService.kt:68]: url is flow still.

**Review by Tom-Ritense — CHANGES_REQUESTED** [review]

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/autoconfiguration/SpotlerConfiguration.kt (hunk 1 above)
Comment (Tom-Ritense) [reply spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/autoconfiguration/SpotlerConfiguration.kt:27]: RestTemplateBuilder()
            .setConnectTimeout(Duration.ofSeconds(10))
            .setReadTimeout(Duration.ofSeconds(10))
            .build()

File: spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/SubmitMessage.kt (hunk 3 above)
Comment (sirolf2009) [reply spotler/backend/plugin/src/main/kotlin/com/ritense/spotler/domain/SubmitMessage.kt:25]: Yeah a flow is a sequence of actions that every email will go through. like set subject, set sender

**Review by Tom-Ritense — APPROVED** [review]

**Review by Tom-Ritense — APPROVED** [review]

**Pull request merged**